myia/utils.py

```diff
@@ -150,19 +200,95 @@ def __call__(self, *data):
         """
         d0 = data[0]
         t = type(d0)
-        if t in self.dispatch:
-            return self.dispatch[t](self, *data)
-        # elif hasattr(d0, '__smap__'):
-        #     return d0.__smap__(self, *data[1:])
-        else:
-            return self.fn(*data)
+        return self.dispatch[t](self, *data)
 
 
 def smap(fn, *args):
     """Map a function recursively over all scalars in a structure."""
     return StructuralMap(fn)(*args)
 
 
+class Event(list):
+    """Simple Event class.
+
+    >>> e = Event('tickle')
+    >>> e.register(lambda event, n: print("hi" * n))
+    <function>
+    >>> e(5)
+    hihihihihi
+
+    Attributes:
+        name: The name of the event.
+        owner: The object defining this event (defaults to None).
+        history: A function that returns a list of previous events, or
+            None if no history exists.
+
+    """
+
+    def __init__(self, name, owner=None, history=None):
+        """Initialize an Event."""
+        self.name = name
+        self.owner = owner
+        self.history = history
```

Comment:
The idea is that if you want to e.g. track a property on all of a graph's nodes, you can listen to `on_add_node` with history on, and it will call it for all existing nodes first, which spares you the trouble of writing that loop yourself. Most of the `Statistics` in `manager.py` would need to use that feature if they were added on a manager that wasn't empty.

Reply:
I get that, but nothing uses it right now.